### PR TITLE
NFT Widget

### PR DESCRIPTION
**Pull request by nwingt**
[Demo entry](https://button.rinkeby.like.co/in/embed/nft/class/likenft10f06wfaql5fxf3g4sy8v57p98lzp7ad92cu34f9aeyhyeklchznsav5npg)

- [x] Fix Tailwind arbitrary value not working
- [x] Show Liker ID data (if any) of the owner address
- [x] Responsive layout (Work on [likecoin-button-sdk](https://github.com/likecoin/likecoin-button-sdk/pull/27))

### CODE REVIEW

File: .circleci/config.yml [hunk 1]
```diff
@@ -3,7 +3,7 @@ jobs:
   test:
     working_directory: ~/likecoin-button
     docker:
-      - image: node:10
+      - image: node:14
```
Comment (williamchong) [inline .circleci/config.yml:6]: Would we want to update firebase-functions to node 14 also?

File: functions/package.json [hunk 2]
```diff
@@ -13,6 +13,8 @@
     "package": "npm run clean && npm run build && npm run deps:copy && npm run deps:install"
   },
   "dependencies": {
+    "@cosmjs/stargate": "^0.28.7",
+    "@likecoin/iscn-js": "^0.2.0-rc.1",
```
Comment (williamchong) [inline functions/package.json:17]: Please use latest non-rc version

File: src/plugins/vue-i18n.js [hunk 3]
```diff
@@ -18,7 +18,9 @@ async function loadLanguageAsync(i18n, lang) {
     i18n.loadedLanguages = [defaultLocale];
   }
   if (!i18n.loadedLanguages.includes(lang)) {
-    const msgs = await import(/* webpackChunkName: "lang-[request]" */ `@/locales/client/${lang}`);
+    // Fix ESLint error in template string
+    // eslint-disable-next-line prefer-template
+    const msgs = await import(/* webpackChunkName: "lang-[request]" */ '@/locales/client/' + lang);
```
Comment (williamchong) [inline src/plugins/vue-i18n.js:23]: What would be the error if we indded use template string?

File: src/pages/in/embed/nft/class/_classId.vue [hunk 4]
```diff
@@ -0,0 +1,190 @@
...
+      return `${LIKER_LAND_URL_BASE}/nft/class/${this.classId}`;
+    },
+    purchaseURL() {
+      return `${APP_LIKE_CO_URL_BASE}/nft/purchase/${encodeURIComponent(
+        this.iscnId
+      )}`;
+    },
+  },
+  async asyncData({ params }) {
+    const { classId } = params;
```
Comment (williamchong) [inline src/pages/in/embed/nft/class/_classId.vue:51]: I think in future likecoin-button use case, we only want user to input `iscn_id` and don't need to care about nft class id
So I imagine the flow would be `iscn_id` as param, fetch `/purchase` API to see if NFT is minted. If minted, the API would also return class id if you need it.
In short term, maybe we would want the nft widget to auto disable if `iscn_id` is not set or nft is not mint; and show this widget if NFT is minted (ie `classId` exists in api) via querying by `iscn_id`

File: src/nuxt.config.js [hunk 5]
```diff
@@ -148,16 +152,13 @@ const nuxtConfig = {
...
-    /* BUG: cannot parallel with extractCSS */
-    // parallel: true,
     babel: {
       presets: ({ isServer }) => [
         [
           '@nuxt/babel-preset-app',
           {
             targets: isServer
-              ? { node: '8.11.1' }
+              ? { node: '10' }
```
Comment (williamchong) [inline src/nuxt.config.js:161]: Should this be 14?

File: src/package.json [hunk 6]
```diff
@@ -5,7 +5,7 @@
   "author": "William Chong",
   "private": true,
   "engines": {
-    "node": "8.x.x"
+    "node": "14"
```
Comment (williamchong) [inline src/package.json:8]: Please update engines in `functions` also

File: src/package.json [hunk 7]
```diff
@@ -17,10 +17,13 @@
     "build-storybook": "build-storybook -c .storybook -o ../docs",
     "lint": "npm run lint:es && npm run lint:style",
     "lint:es": "eslint --ext .js,.vue .",
+    "lint:es:fix": "eslint --fix --ext .js,.vue .",
     "lint:style": "stylelint -i .gitignore '**/*.{scss,vue}'",
     "precommit": "npm run lint"
   },
   "dependencies": {
+    "@cosmjs/stargate": "^0.28.7",
+    "@likecoin/iscn-js": "^0.2.0-rc.1",
```
Comment (williamchong) [inline src/package.json:26]: Please use non-rc version

File: src/package.json [hunk 8]
```diff
@@ -46,31 +49,35 @@
...
     "eslint-plugin-import": "^2.16.0",
+    "eslint-plugin-jest": "^22.3.0",
+    "eslint-plugin-node": ">=7.0.1",
     "eslint-plugin-nuxt": "^0.4.2",
+    "eslint-plugin-promise": ">=4.0.1",
+    "eslint-plugin-standard": ">=4.0.0",
     "eslint-plugin-vue": "^5.2.2",
-    "extract-css-chunks-webpack-plugin": "^4.8.0",
-    "node-sass": "^4.11.0",
+    "node-sass": "^4.14.0",
```
Comment (williamchong) [inline src/package.json:73]: I think we want >=5 or 6 for better M1 dev exp, I don't dare merging the pr in liker-land yet, maybe you can try it here?
https://github.com/likecoin/liker-land/pull/523

File: src/util/nft.js [hunk 9]
```diff
@@ -0,0 +1,28 @@
+// eslint-disable-next-line import/no-extraneous-dependencies
+import { ISCNQueryClient, ISCNSigningClient } from '@likecoin/iscn-js';
```
Comment (williamchong) [inline src/util/nft.js:2]: Why is iscn-js extraneous-dependencies ?

File: src/pages/in/embed/nft/class/_classId.vue (hunk 4 above)
Comment (williamchong) [reply src/pages/in/embed/nft/class/_classId.vue:51]: Or would this page be a used in a separated use case where we put int `class_id` only? Where would we use this?

File: src/pages/in/embed/nft/class/_classId.vue [hunk 10]
```diff
@@ -0,0 +1,190 @@
...
+    const [
+      chainMetadata = {},
+      apiMetadataResult,
+      ownersResult,
+      purchaseInfoResult,
+    ] = await Promise.all([
+      getClassInfo(classId),
+      apiGetNFTMetadata({ classId })
+        // eslint-disable-next-line no-console
+        .catch(error => console.error(error)),
```
Comment (williamchong) [inline src/pages/in/embed/nft/class/_classId.vue:62]: Just as a note: if we want to support NFT other than writing NFT in the future, metadata should be fetched from `uri` field of the on-chain metadata

File: src/pages/in/embed/nft/class/_classId.vue [hunk 11]
```diff
@@ -0,0 +1,190 @@
...
+    const ownersMap = (ownersResult || {}).data || {};
+    const nftCollectedCount = Object.values(ownersMap).reduce(
+      (count, collected) => count + collected.length,
+      0
+    );
+    const nftCollectorCount = Object.keys(ownersMap).length;
+
+    const likerDataResult = await apiGetLikerDataByAddress(iscnOwnerAddress)
+      // eslint-disable-next-line no-console
+      .catch(error => console.error(error));
```
Comment (williamchong) [inline src/pages/in/embed/nft/class/_classId.vue:112]: You can save a `|| {}` guard below if you return `{}` here, not sure which one is cleaner

**Review by williamchong — CHANGES_REQUESTED** [review]
Not sure if this widget is supposed to have iscn_id or class_id as key
I suppose we very much want to use iscn_id to reduce number of concepts

File: src/plugins/vue-i18n.js (hunk 3 above)
Comment (nwingt) [reply src/plugins/vue-i18n.js:23]:
```
TypeError: Cannot read property 'range' of null
Occurred while linting /Users/nwingt/Repos/likecoin-button/src/plugins/vue-i18n.js:22
    at SourceCode.getTokenBefore (/Users/nwingt/Repos/likecoin-button/src/node_modules/eslint/lib/token-store/index.js:303:18)
    at checkSpacingBefore (/Users/nwingt/Repos/likecoin-button/src/node_modules/eslint/lib/rules/template-curly-spacing.js:60:42)
    at TemplateElement (/Users/nwingt/Repos/likecoin-button/src/node_modules/eslint/lib/rules/template-curly-spacing.js:119:17)
    at /Users/nwingt/Repos/likecoin-button/src/node_modules/eslint/lib/util/safe-emitter.js:45:58
    at Array.forEach (<anonymous>)
    at Object.emit (/Users/nwingt/Repos/likecoin-button/src/node_modules/eslint/lib/util/safe-emitter.js:45:38)
    at NodeEventGenerator.applySelector (/Users/nwingt/Repos/likecoin-button/src/node_modules/eslint/lib/util/node-event-generator.js:251:26)
    at NodeEventGenerator.applySelectors (/Users/nwingt/Repos/likecoin-button/src/node_modules/eslint/lib/util/node-event-generator.js:280:22)
    at NodeEventGenerator.enterNode (/Users/nwingt/Repos/likecoin-button/src/node_modules/eslint/lib/util/node-event-generator.js:294:14)
    at CodePathAnalyzer.enterNode (/Users/nwingt/Repos/likecoin-button/src/node_modules/eslint/lib/code-path-analysis/code-path-analyzer.js:632:23)
```

File: .circleci/config.yml (hunk 1 above)
Comment (nwingt) [reply .circleci/config.yml:6]: Do you mean the `firebase-functions` [package](https://www.npmjs.com/package/firebase-functions)?

Comment (williamchong) [reply .circleci/config.yml:6]: i mean the engines in functions package json, which control the node version of deployed functions

File: src/components/LikeButtonV2/LikeButtonV2.badge.vue [hunk 12]
```diff
@@ -35,13 +35,13 @@
           </transition>
         </g>
         <transition
+          :css="false"
+          mode="in-out"
           @before-appear="overlayBeforeEnter"
           @apper="overlayEnter"
```
Comment (AuroraHuang22) [inline src/components/LikeButtonV2/LikeButtonV2.badge.vue:41]:
```suggestion
          @appear="overlayEnter"
```

**Review by AuroraHuang22 — APPROVED** [review]

**Pull request merged**